proposals/0005-fallback-rounting.md

```diff
@@ -0,0 +1,171 @@
+            model: gpt-3.5-turbo
+            api_key: ${env:OPENAI_API_KEY}
+            default_params:
+              temperature: 0
+```
+
+Notes:
+- The bare minimal setup must include one pool with at least one model.
+  We don't require to set up more than one model to simplify gateway setup for users who just want to try it or experiment.
+  That said, we may want to show a warning that such a pool better have some redundancy to be more resilient to provider outages.
+- Each pool ID should have a unique name across other pools of this type (e.g. language pools)
+- Each model ID should be unique for the pool it's defined in
+- Pool and model definition include the "enabled" field to simplify disabling pools/models without a need to remove their definitions from the file
+
+### Pool API
+
+This could be useful to have an ability to inspect the current pool setup in case Glide users don't have access to the underlying gateway configuration (it's likely that gateway is deployed by the team that doesn't do LLM work directly, but rather helps with operations/engineering).
+
+In that case, Glide provides the following API (see GEP0002, for other API):
+
+```
+GET /v1/language/
+# This will return all enabled pools like:
+[
+  {
+    "id": "mypool",
+    "strategy": "priority",
+    "models": [
+      {
+        "id": "openai-boring",
+        "openai": {
+          "model": "gpt-3.5-turbo",
+          "api_key": "[REDACTED]",
+          "default_params": {
+            "temperature": 0
+          }
+        }
+      }
+    ]
+  }
+]
+```
+
+API keys and any other sensitive data is going to be redacted in this output.
+
+### Testing
+
+The real providers are costly to be used for E2E testing which is useful thing 
+to have implemented for Glide to ensure as few regressions as possible. 
+
+In addition to that, real providers may not be convenient to use to check for abnormal conditions like rate limiting.
+
+Hence, we need a fake provider that will help to emulate needed conditions for us:
+
+- In-memory provider that is used for unit/integration tests
+- A client and a simple server that would generate needed faulty conditions in E2E tests according to a given config. Let's call it `FaultyProvider`.
+
+The FaultyProvider could be useful in other cases that resemble E2E scenario like 
+- load testing e.g. sending a lot of requests through Glide to measure its performance
+- or resiliency testing e.g. send a lot of requests through Glide to a pool with a few FaultyProvider configured to fail differently and measure how many requests end up failing
+
+## Model Pool Use Cases
+
+The model pool abstraction seems like a powerful idea that allows to do these things (not an exhaustive list):
+
+- Switch clients to a new pool with different setup in a backward-compatible manner (e.g. keeps the old pool in place, add a new one, switch the app to the new pool, remove the old one)
```

Comment:
Can this be done in an interactive manner? I.e. 90% of traffic to old pool and 10% to the new pool then 50%/50%....and so on until the old pool is no longer used?

Reply:
@mkrueger12 I thought about that. It's like blue-green/canary deployment. However, not sure if this is possible to achieve via Glide alone. The reason is that clients controls the router ID to which they send traffic. If we want to switch traffic between two routers, that change should happen on the client side. With that, it's still possible to achieve but on the application deployment level.

If users deploy their apps via blue-green deployment, then at some point they may have two versions of applications: one that routes traffic to old router (the old release) + one that routes traffic to new one (the new release). As they gradually spin down the old release, the portion of traffic that comes to the old router naturally decrease and vice versa.

Reply:
Ah interesting. I think this if fine. I will add something in the docs that basically states this.